NeoML/src/TraditionalML/ProblemWrappers.cpp

```diff
@@ -209,86 +137,48 @@ CProblemNotNullWeightsView::CProblemNotNullWeightsView( const IProblem* _inner )
 
 	int originalVectorCount = inner->GetVectorCount();
 	if( originalVectorCount > 0 ) {
-		// fill nullWeightElementsMap, and adjust local MatrixDesc
-		viewMatrixDesc = _inner->GetMatrix();
-		// we are going to remap some elements, so create our own arrays of pointers
-		viewMatrixDesc.PointerB = static_cast<int*>( 
-			ALLOCATE_MEMORY( CurrentMemoryManager, originalVectorCount * sizeof( int ) ) );
-		viewMatrixDesc.PointerE = static_cast<int*>( 
-			ALLOCATE_MEMORY( CurrentMemoryManager, originalVectorCount * sizeof( int ) ) );
-
-		int nullWeightElmentsCount = 0;
-		for( int i = 0; i < originalVectorCount - nullWeightElmentsCount; ) {
-			int iScanned = i + nullWeightElmentsCount;
+		int nullWeightElementsCount = 0;
+		for( int i = 0; i < originalVectorCount - nullWeightElementsCount; ) {
+			int iScanned = i + nullWeightElementsCount;
 			if( _inner->GetVectorWeight( iScanned ) == 0 ) {
-				++nullWeightElmentsCount;
+				++nullWeightElementsCount;
 			} else {
-				viewMatrixDesc.PointerB[i] = _inner->GetMatrix().PointerB[iScanned];
-				viewMatrixDesc.PointerE[i] = _inner->GetMatrix().PointerE[iScanned];
 				notNullWeightElementsIndices.Add( iScanned );
 				++i;
 			}
 		}
-		viewMatrixDesc.Height -= nullWeightElmentsCount;
+
+		// fill nullWeightElementsMap, and then adjust local MatrixDesc if needed
+		viewMatrixDesc = _inner->GetMatrix();
+		viewMatrixDesc.Height -= nullWeightElementsCount;
+		NeoAssert( viewMatrixDesc.Height == notNullWeightElementsIndices.Size() );
+
+		if( nullWeightElementsCount != 0 && viewMatrixDesc.Height > 0 ) {
+			// we are going to remap some elements, so let's create our own arrays of pointers
+			viewMatrixDesc.PointerB = static_cast<int*>(
+				ALLOCATE_MEMORY( CurrentMemoryManager, viewMatrixDesc.Height * sizeof( int ) ) );
+			viewMatrixDesc.PointerE = static_cast<int*>(
+				ALLOCATE_MEMORY( CurrentMemoryManager, viewMatrixDesc.Height * sizeof( int ) ) );
+
+			for( int i = 0; i < viewMatrixDesc.Height; ++i ) {
+				int originalIndex = notNullWeightElementsIndices[i];
```

Comment:
We can form array `notNullWeightElementsIndices` in this cycle, in order to avoid any allocations if `nullWeightElementsCount == 0`. Also before this cycle `notNullWeightElementsIndices.SetBufferSize( viewMatrixDesc.Height )`.

Reply:
Ok, done